### DISCUSSION TITLE
Optimize NPM package manager with dedupe option

### DISCUSSION THREAD

**Discussion by jelhan** [renovatebot/renovate#25455] [Suggest an Idea]
### Tell us more.

NPM package manager with `npmDedupe` option is not very performant. I run into timeouts with my project. When looking into why it takes so much longer with deduplication, I noticed that the implementation could be improved.

Currently it does the following if `npmDedupe` is in `postUpdateOptions`:

1. Run `npm install` _without_ the `--package-lock-only` flag. This causes all packages to be downloaded and a `node_modules/` folder being created. Very time consuming. https://github.com/renovatebot/renovate/blob/3df10279cd3422c0ff815f37d8d6870105929462/lib/modules/manager/npm/post-update/npm.ts#L54-L63
2. Run `npm dedupe` as a second step. https://github.com/renovatebot/renovate/blob/3df10279cd3422c0ff815f37d8d6870105929462/lib/modules/manager/npm/post-update/npm.ts#L132-L136

Two optimizations are possible:

1. `npm install` can deduplicate the dependencies directly with the `--prefer-dedupe` flag. Instead of `npm install --no-audit && npm dedupe` we can do `npm install --prefer-dedupe`, which is much faster.
2. No need to download and install the packages for deduplication. `--prefer-dedupe` and `--package-lock-only` flags are working well together.

The performance difference is huge. `npm install --prefer-dedupe --package-lock-only --no-audit` is finished within 1 second on my test project while `npm install --no-audit && npm dedupe` takes minutes.

_Note: Converted from bug report in #25454 to this feature request._

**Comment by viceice** [comment]
sounds good, which npm version introduced `prefer-dedupe`? we probably need to check it before we can use the new code path.

**Reply by jelhan** [reply]
The `--prefer-dedupe` flag has been added in NPM 7. ([Source](https://docs.npmjs.com/cli/v7/using-npm/changelog#npm-dedupe))

NPM 7 was shipped with node 15. Node 15 was released in October 2020. Node 14, the last version with NPM 6, reached end of life at April 30 this year. ([Source](https://endoflife.date/nodejs))

NPM only supports the latest version. Additionally the project "will try to backport - when possible - security patches to versions of npm currently shipping with 'maintained' Node.js versions." ([Source](https://github.com/npm/cli/wiki/Support-Policy#security-issues--backports))

To summarize: `--prefer-dedupe` flag is available in all NPM versions, which could be considered as maintained.

**Reply by rarkins** [reply]
I think we can say that Renovate's npmDedupe option is only supported for Node 7+

**Comment by rarkins** [comment]
Converted to feature request: #25462
